src/x.c

```diff
@@ -440,27 +511,39 @@ void x_set_picture_clip_region(xcb_connection_t *c, xcb_render_picture_t pict,
 		};
 	}
 
-	xcb_generic_error_t *e = xcb_request_check(
-	    c, xcb_render_set_picture_clip_rectangles_checked(
-	           c, pict, clip_x_origin, clip_y_origin, to_u32_checked(nrects), xrects));
+	xcb_generic_error_t *e =
+	    xcb_request_check(c->c, xcb_render_set_picture_clip_rectangles_checked(
+	                                c->c, pict, clip_x_origin, clip_y_origin,
+	                                to_u32_checked(nrects), xrects));
 	if (e) {
 		log_error_x_error(e, "Failed to set clip region");
 		free(e);
 	}
 	free(xrects);
 }
 
-void x_clear_picture_clip_region(xcb_connection_t *c, xcb_render_picture_t pict) {
+void x_clear_picture_clip_region(struct x_connection *c, xcb_render_picture_t pict) {
 	assert(pict != XCB_NONE);
 	xcb_render_change_picture_value_list_t v = {.clipmask = XCB_NONE};
 	xcb_generic_error_t *e = xcb_request_check(
-	    c, xcb_render_change_picture_checked(c, pict, XCB_RENDER_CP_CLIP_MASK, &v));
+	    c->c, xcb_render_change_picture_checked(c->c, pict, XCB_RENDER_CP_CLIP_MASK, &v));
 	if (e) {
 		log_error_x_error(e, "failed to clear clip region");
 		free(e);
 	}
 }
 
+/**
+ * Destroy a <code>Picture</code>.
+ *
+ * Picture must be valid.
+ */
+void x_free_picture(struct x_connection *c, xcb_render_picture_t p) {
+	assert(p != XCB_NONE);
+	auto cookie = xcb_render_free_picture(c->c, p);
+	set_cant_fail_cookie(c, cookie);
+}
```

Comment:
this makes sure we don't double free a picture.

Reply:
it should be the x server's headache, not ours. furthermore, double freeing a picture pretty much results in an x error and if you want to trace where it happens there is [xrescheck](https://github.com/absolutelynothelix/xrescheck).

Reply:
i don't quite understand what you are trying to say? `set_cant_fail_cookie` is what tracks the X error.

Reply:
calling the `xcb_render_free_picture` function pretty much sends a corresponding request to the x server and handling a possible double free is up to the x server. for us it's not something horrible like an actual double free in terms of memory allocated by us, it's just an x error. and i can't really imagine a scenario where something goes horribly wrong because of a double freed picture.

> `set_cant_fail_cookie` is what tracks the X error.

as i can see from the code, it considers the error unrecoverable and aborts.

basically, my point is that personally i wouldn't care if a picture is successfully freed or not because in most cases it'll result in a harmless error that could be traced with xrescheck. if you disagree with me i won't mind such mechanism but it's still unclear when to prefer it over a plain `xcb_render_free_picture` call because it doesn't seem to be always used.

Reply:
because if we free a picture twice, it means there is a bug in our code. it would be nice if we can catch it.

does that make sense?

Reply:
these errors are not that severe to abort on them and could be caught without this mechanism but... nevermind. i'd like to see other nitpicks addressed and i'll merge this pull request.

Reply:
well there is also `set_debug_cant_fail_cookie`